### PR TITLE
torch RaggedTensor class and batch support for SparseConv

### PR DESCRIPTION
**Pull request by sanskar107**
Added a container for RaggedTensor for torch and enable batch support in `fixed_radius_search`, `SparseConv` and `SparseConvTranspose`. Also verified that gradients flow is working.
Current class has some limitations as PyTorch doesn't support binding static methods and read/writable data members of a class.
1) Unable to use `isinstance` to check for RaggedTensor object, as it always returns torchscript object. Hence using `str(points)[:12] == "RaggedTensor"` for the same.
2) Cannot access properties like `r_tensor.dtype`, `r_tensor.device`.

One workaround to improvise is to define a python wrapper `RaggedTensor` class in `python/open3d/ml/torch/classes/` and within that class, we can use this binded C++ class and add some extra properties. This would also solve the dostring issue.

**UPDATE**
Added python wrapper to overcome above limitations.

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572)

### CODE REVIEW

**Review by benjaminum — CHANGES_REQUESTED** [review]
Seems like adding a python wrapper is the only way to workaround the limitations for now.

Can you add some simple python unit tests that test creation and backprop (maybe for addition of two ragged tensors)

> Reviewed 3 of 7 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572#-:-McD6oaV4e0lsW--UNIZ:b-dx0cib)* status: 4 of 7 files reviewed, 2 unresolved discussions (waiting on @sanskar107)

___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h, line 49 at r2](https://reviewable.io/reviews/intel-isl/open3d/3572#-McD1bV67b1JXnFVpGAQ:-McD1bV76C8wDaALf3uV:bpm255m) ([raw file](https://github.com/intel-isl/open3d/blob/7a1e2f02092a7a0e0a8662b3bbd2c1f867d8361a/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h#L49)):*
> ```objc
>     ///
>     /// The returned `RaggedTensor` corresponds with the python list defined by:
>     /// ```python
> ```

Does this work in doxygen or should it be```\code{.py} ... \endcode```
___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp, line 33 at r2](https://reviewable.io/reviews/intel-isl/open3d/3572#-McD3EDDAlmmuFXEBuFz:-McD3EDDAlmmuFXEBuG-:brcc1n2) ([raw file](https://github.com/intel-isl/open3d/blob/7a1e2f02092a7a0e0a8662b3bbd2c1f867d8361a/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp#L33)):*
> ```cpp
> 
> c10::intrusive_ptr<RaggedTensor> RaggedTensor::FromRowSplits(
>         torch::Tensor values, torch::Tensor row_splits) const {
> ```

Tensorflow has a ```validate=True``` argument. Can we add this argument and skip the expensive "monotonically increasing" check for validate=False?

**Review by sanskar107 — COMMENTED** [review]
I've added python wrapper and unit tests.

> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572#-:-McOaVX3-PNlBM5BhfNG:b-7172ir)* status: 2 of 11 files reviewed, 2 unresolved discussions (waiting on @benjaminum)

___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h, line 49 at r2](https://reviewable.io/reviews/intel-isl/open3d/3572#-McD1bV67b1JXnFVpGAQ:-McOaavc6v3dim8PcFU5:b-ha5j47) ([raw file](https://github.com/intel-isl/open3d/blob/7a1e2f02092a7a0e0a8662b3bbd2c1f867d8361a/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h#L49)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Does this work in doxygen or should it be```\code{.py} ... \endcode```
</blockquote></details>

I took this format from our `core::Tensor` which seems to look fine in the C++ doxygen documentation.
___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp, line 33 at r2](https://reviewable.io/reviews/intel-isl/open3d/3572#-McD3EDDAlmmuFXEBuFz:-McOao8dBiOODQZYtf4H:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/7a1e2f02092a7a0e0a8662b3bbd2c1f867d8361a/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp#L33)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Tensorflow has a ```validate=True``` argument. Can we add this argument and skip the expensive "monotonically increasing" check for validate=False?
</blockquote></details>

Done.

**Review by benjaminum — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572)* status: 1 of 11 files reviewed, 7 unresolved discussions (waiting on @benjaminum and @sanskar107)

___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp, line 80 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTL32e6RmiUsUA-d4h:-McTL32e6RmiUsUA-d4i:b-2ccyd1) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp#L80)):*
> ```cpp
> }
> 
> c10::intrusive_ptr<RaggedTensor> RaggedTensor::Concat(
> ```

Do we use this function at the moment? It seems unfinished and is not exposed in python. We can deactivate it for now (throw unimplemented exception)
___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 182 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTMXQm4w6Tx_nuBZVD:-McTMXQm4w6Tx_nuBZVE:b-b91gax) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/open3d/ml/torch/classes/ragged_tensor.py#L182)):*
> ```Python
>         return self
> 
>     def __div__(self, other):
> ```

do we need __div__? I don't see it in the python3 docs
___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 189 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTNGSpAjm2yyB--pwP:-McTNGSpAjm2yyB--pwQ:bvno60x) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/open3d/ml/torch/classes/ragged_tensor.py#L189)):*
> ```Python
>         return self
> 
>     def __truediv__(self, other):
> ```

floordiv and truediv have the same implementation
___
*[python/open3d/ml/torch/python/layers/convolutions.py, line 756 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTNVvN6WkgBw0DNQKd:-McTNVvN6WkgBw0DNQKe:b-u5htow) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/open3d/ml/torch/python/layers/convolutions.py#L756)):*
> ```Python
> 
>         if out_positions_split is not None:
>             out_features = classes.RaggedTensor.from_row_splits(
> ```

we can make use of the shallow copy for torch
___
*[python/test/ml_ops/test_ragged_tensor.py, line 42 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTPk9f3_yFV_C-eTOg:-McTPk9f3_yFV_C-eTOh:bb3bs1n) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/test/ml_ops/test_ragged_tensor.py#L42)):*
> ```Python
> 
> # this op is only available for torch
> ```

Please add tests for the operators. Maybe copy and adapt from here
https://github.com/intel-isl/Open3D/blob/master/python/test/core/test_core.py#L347

**Review by sanskar107 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572)* status: 1 of 11 files reviewed, 7 unresolved discussions (waiting on @benjaminum)

___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp, line 80 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTL32e6RmiUsUA-d4h:-McUbO757LA4L8oY6Pu-:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.cpp#L80)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Do we use this function at the moment? It seems unfinished and is not exposed in python. We can deactivate it for now (throw unimplemented exception)
</blockquote></details>

Done.
___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 182 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTMXQm4w6Tx_nuBZVD:-McUbOy88zEt9pdaekHx:b-42azex) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/open3d/ml/torch/classes/ragged_tensor.py#L182)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

do we need __div__? I don't see it in the python3 docs
</blockquote></details>

Removed `div` and `idiv`
___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 189 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTNGSpAjm2yyB--pwP:-McUbTbk91Tb6_uxCWcw:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/open3d/ml/torch/classes/ragged_tensor.py#L189)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

floordiv and truediv have the same implementation
</blockquote></details>

Done.
___
*[python/open3d/ml/torch/python/layers/convolutions.py, line 756 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTNVvN6WkgBw0DNQKd:-McUbUUiAgs-U9iCfBMn:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/open3d/ml/torch/python/layers/convolutions.py#L756)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

we can make use of the shallow copy for torch
</blockquote></details>

Done.
___
*[python/test/ml_ops/test_ragged_tensor.py, line 42 at r3](https://reviewable.io/reviews/intel-isl/open3d/3572#-McTPk9f3_yFV_C-eTOg:-McUbV536YwuVhDDJlEF:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/86c14125b0ae871ea2501f12d8c7c5a214739dd9/python/test/ml_ops/test_ragged_tensor.py#L42)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Please add tests for the operators. Maybe copy and adapt from here
https://github.com/intel-isl/Open3D/blob/master/python/test/core/test_core.py#L347

</blockquote></details>

Done.

**Review by benjaminum — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572)* status: 1 of 11 files reviewed, 9 unresolved discussions (waiting on @benjaminum and @sanskar107)

___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h, line 85 at r4](https://reviewable.io/reviews/intel-isl/open3d/3572#-Mcc9JtoDNBNdD4ZEe36:-Mcc9JtoDNBNdD4ZEe37:b-mxb4fd) ([raw file](https://github.com/intel-isl/open3d/blob/a6d675ccdea57fb850f34de1c6a223e501ce259e/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h#L85)):*
> ```objc
> 
>     template <typename T>
>     c10::intrusive_ptr<RaggedTensor> Add(T value) const {
> ```

If both operand are ragged tensors the element-wise operations will work even for incompatible  tensors. For instance `[ [1], [2,3] ] + [ [1,2], [3] ] = [ [2], [4,6] ]` will work but should throw an exception.
Seems like we need a specialized function for RaggedTensor args.
___
*[python/test/ml_ops/test_ragged_tensor.py, line 43 at r4](https://reviewable.io/reviews/intel-isl/open3d/3572#-MccAeWA5VPkUo9Gl_AC:-MccAeWA5VPkUo9Gl_AD:b-pi9qd3) ([raw file](https://github.com/intel-isl/open3d/blob/a6d675ccdea57fb850f34de1c6a223e501ce259e/python/test/ml_ops/test_ragged_tensor.py#L43)):*
> ```Python
> # this op is only available for torch
> ```

Please add tests checking some failure cases with incompatible tensors.

**Review by sanskar107 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572)* status: 1 of 11 files reviewed, 9 unresolved discussions (waiting on @benjaminum)

___
*[cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h, line 85 at r4](https://reviewable.io/reviews/intel-isl/open3d/3572#-Mcc9JtoDNBNdD4ZEe36:-McmqhKoDCu-6M7Cy_rO:b-ag11jm) ([raw file](https://github.com/intel-isl/open3d/blob/a6d675ccdea57fb850f34de1c6a223e501ce259e/cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h#L85)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

If both operand are ragged tensors the element-wise operations will work even for incompatible  tensors. For instance `[ [1], [2,3] ] + [ [1,2], [3] ] = [ [2], [4,6] ]` will work but should throw an exception.
Seems like we need a specialized function for RaggedTensor args.

</blockquote></details>

I meant to add these binary ops only for `torch::Tensor` and Scalar values in C++ API and add support for RaggedTensor through Python wrapper. I've added shape checking for incompatible RaggedTensors in python wrapper. We can add everything in C++ API once we migrate to torch 1.9.
___
*[python/test/ml_ops/test_ragged_tensor.py, line 43 at r4](https://reviewable.io/reviews/intel-isl/open3d/3572#-MccAeWA5VPkUo9Gl_AC:-McmsQp98VU46VhtDQW8:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/a6d675ccdea57fb850f34de1c6a223e501ce259e/python/test/ml_ops/test_ragged_tensor.py#L43)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Please add tests checking some failure cases with incompatible tensors.
</blockquote></details>

Done.

**Review by benjaminum — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572)* status: 1 of 11 files reviewed, 11 unresolved discussions (waiting on @benjaminum and @sanskar107)

___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 196 at r5](https://reviewable.io/reviews/intel-isl/open3d/3572#-McnhtWx6KFISqX48-_3:-McnhtWx6KFISqX48-_4:b541c2c) ([raw file](https://github.com/intel-isl/open3d/blob/28035017ea49d52c152aad9ae7d359030112fbd0/python/open3d/ml/torch/classes/ragged_tensor.py#L196)):*
> ```Python
>         return self
> 
>     def convert_to_tensor(self, value):
> ```

Just a suggestion, maybe hide this function it seems to be meant for internal use.
___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 208 at r5](https://reviewable.io/reviews/intel-isl/open3d/3572#-Mcni0KnAysTIxwxt0qh:-Mcni0KnAysTIxwxt0qi:b-9rzjo8) ([raw file](https://github.com/intel-isl/open3d/blob/28035017ea49d52c152aad9ae7d359030112fbd0/python/open3d/ml/torch/classes/ragged_tensor.py#L208)):*
> ```Python
>             return value
>         elif isinstance(value, (int, float, bool)):
>             return torch.full(self.values.shape, value)
> ```

Is there no broadcasting on the cpp side?

**Review by sanskar107 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572)* status: 1 of 11 files reviewed, 11 unresolved discussions (waiting on @benjaminum)

___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 196 at r5](https://reviewable.io/reviews/intel-isl/open3d/3572#-McnhtWx6KFISqX48-_3:-McnpOS09YKe8yeL3fO0:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/28035017ea49d52c152aad9ae7d359030112fbd0/python/open3d/ml/torch/classes/ragged_tensor.py#L196)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Just a suggestion, maybe hide this function it seems to be meant for internal use.
</blockquote></details>

Done.
___
*[python/open3d/ml/torch/classes/ragged_tensor.py, line 208 at r5](https://reviewable.io/reviews/intel-isl/open3d/3572#-Mcni0KnAysTIxwxt0qh:-McnpP3UDeSSpUwcMwWQ:bnlp9wa) ([raw file](https://github.com/intel-isl/open3d/blob/28035017ea49d52c152aad9ae7d359030112fbd0/python/open3d/ml/torch/classes/ragged_tensor.py#L208)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

Is there no broadcasting on the cpp side?
</blockquote></details>

Yes, there is. Modified it.

**Review by benjaminum — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3572#-:-McnvT6G2SFulOaLRejt:bnfp4nl)* status: 1 of 11 files reviewed, 11 unresolved discussions (waiting on @benjaminum)

File: cpp/open3d/ml/pytorch/CMakeLists.txt [hunk 1]
```diff
@@ -56,6 +56,9 @@ target_sources(open3d_torch_ops PRIVATE
     misc/VoxelPoolingOps.cpp
 )
 
+target_sources(open3d_torch_ops PRIVATE
+ragged_tensor/RaggedTensor.cpp)
```
Comment (yxlao) [inline cpp/open3d/ml/pytorch/CMakeLists.txt:60]: indentation

File: cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+//
+
+#include <vector>
+
+#include "open3d/ml/pytorch/TorchHelper.h"
+#include "torch/custom_class.h"
```
Comment (yxlao) [inline cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h:31]: `<>` for 3rd-party headers

File: cpp/open3d/ml/pytorch/CMakeLists.txt (hunk 1 above)
Comment (sanskar107) [reply cpp/open3d/ml/pytorch/CMakeLists.txt]: Done

File: cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h (hunk 2 above)
Comment (sanskar107) [reply cpp/open3d/ml/pytorch/ragged_tensor/RaggedTensor.h]: Done

**Pull request merged**